
<!doctype html>
<html lang="en" class="no-js">
  <head>

      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">

        <meta name="description" content="Uniform Manifold Approximation and Projection">


        <meta name="author" content="Leland McInnes">


        <link rel="canonical" href="https://lmcinnes.github.io/umap/">



        <link rel="next" href="basic_usage/">


      <link rel="icon" href="doc/logo_large.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.46">



        <title>UMAP</title>



      <link rel="stylesheet" href="assets/stylesheets/main.6f8fc17f.min.css">


        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">












        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>



    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>






  </head>









    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">


    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">


        <a href="#umap" class="md-skip">
          Skip to content
        </a>

    </div>
    <div data-md-component="announce">

    </div>






<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="UMAP" class="md-header__button md-logo" aria-label="UMAP" data-md-component="logo">

  <img src="doc/logo_large.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">

      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            UMAP
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">

              Home

          </span>
        </div>
      </div>
    </div>


        <form class="md-header__option" data-md-component="palette">




    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">

      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>





    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">

      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>


</form>



      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>



      <label class="md-header__button md-icon" for="__search">

        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">

        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>

        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">

        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">

          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>

        <div class="md-search__suggest" data-md-component="search-suggest"></div>

    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>


      <div class="md-header__source">
        <a href="https://github.com/lmcinnes/umap" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">

    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    lmcinnes/umap
  </div>
</a>
      </div>

  </nav>

</header>

    <div class="md-container" data-md-component="container">






      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">



              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="UMAP" class="md-nav__button md-logo" aria-label="UMAP" data-md-component="logo">

  <img src="doc/logo_large.png" alt="logo">

    </a>
    UMAP
  </label>

    <div class="md-nav__source">
      <a href="https://github.com/lmcinnes/umap" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">

    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    lmcinnes/umap
  </div>
</a>
    </div>

  <ul class="md-nav__list" data-md-scrollfix>









    <li class="md-nav__item md-nav__item--active">

      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">





        <label class="md-nav__link md-nav__link--active" for="__toc">


  <span class="md-ellipsis">
    Home
  </span>


          <span class="md-nav__icon md-icon"></span>
        </label>

      <a href="." class="md-nav__link md-nav__link--active">


  <span class="md-ellipsis">
    Home
  </span>


      </a>



<nav class="md-nav md-nav--secondary" aria-label="Table of contents">






    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>

        <li class="md-nav__item">
  <a href="#installing" class="md-nav__link">
    <span class="md-ellipsis">
      Installing
    </span>
  </a>

    <nav class="md-nav" aria-label="Installing">
      <ul class="md-nav__list">

          <li class="md-nav__item">
  <a href="#install-options" class="md-nav__link">
    <span class="md-ellipsis">
      Install Options
    </span>
  </a>

</li>

          <li class="md-nav__item">
  <a href="#manual-development-install" class="md-nav__link">
    <span class="md-ellipsis">
      Manual Development Install
    </span>
  </a>

</li>

          <li class="md-nav__item">
  <a href="#development-testing-with-testmon" class="md-nav__link">
    <span class="md-ellipsis">
      Development Testing with testmon
    </span>
  </a>

</li>

      </ul>
    </nav>

</li>

        <li class="md-nav__item">
  <a href="#how-to-use-umap" class="md-nav__link">
    <span class="md-ellipsis">
      How to use UMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#benefits-of-umap" class="md-nav__link">
    <span class="md-ellipsis">
      Benefits of UMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#performance-and-examples" class="md-nav__link">
    <span class="md-ellipsis">
      Performance and Examples
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#plotting" class="md-nav__link">
    <span class="md-ellipsis">
      Plotting
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#parametric-umap" class="md-nav__link">
    <span class="md-ellipsis">
      Parametric UMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#densmap" class="md-nav__link">
    <span class="md-ellipsis">
      densMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#interactive-umap-with-nomic-atlas" class="md-nav__link">
    <span class="md-ellipsis">
      Interactive UMAP with Nomic Atlas
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#gpu-accelerated-umap-with-torchdr" class="md-nav__link">
    <span class="md-ellipsis">
      GPU-Accelerated UMAP with torchdr
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#help-and-support" class="md-nav__link">
    <span class="md-ellipsis">
      Help and Support
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      Citation
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    <span class="md-ellipsis">
      License
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#contributing" class="md-nav__link">
    <span class="md-ellipsis">
      Contributing
    </span>
  </a>

</li>

    </ul>

</nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >


          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">


  <span class="md-ellipsis">
    Getting Started
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="basic_usage/" class="md-nav__link">


  <span class="md-ellipsis">
    Installation
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="how_umap_works/" class="md-nav__link">


  <span class="md-ellipsis">
    How UMAP Works
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="reproducibility/" class="md-nav__link">


  <span class="md-ellipsis">
    Reproducibility
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >


          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">


  <span class="md-ellipsis">
    User Guide
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="basic_usage/" class="md-nav__link">


  <span class="md-ellipsis">
    Installation
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="parameters/" class="md-nav__link">


  <span class="md-ellipsis">
    Parameters
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="supervised/" class="md-nav__link">


  <span class="md-ellipsis">
    Supervised Dimensionality Reduction
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="clustering/" class="md-nav__link">


  <span class="md-ellipsis">
    Clustering
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="inverse_transform/" class="md-nav__link">


  <span class="md-ellipsis">
    Inverse Transform
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="sparse/" class="md-nav__link">


  <span class="md-ellipsis">
    Sparse Data
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >


          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">


  <span class="md-ellipsis">
    Advanced Topics
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Advanced Topics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="composing_models/" class="md-nav__link">


  <span class="md-ellipsis">
    Composing Models
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="aligned_umap_basic_usage/" class="md-nav__link">


  <span class="md-ellipsis">
    Aligned UMAP
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="aligned_umap_politics_demo/" class="md-nav__link">


  <span class="md-ellipsis">
    Aligned UMAP - Politics Demo
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="embedding_space/" class="md-nav__link">


  <span class="md-ellipsis">
    Embedding Space
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="exploratory_analysis/" class="md-nav__link">


  <span class="md-ellipsis">
    Exploratory Analysis
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >


          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">


  <span class="md-ellipsis">
    Variants and Extensions
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Variants and Extensions
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="densmap_demo/" class="md-nav__link">


  <span class="md-ellipsis">
    densMAP
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="parametric_umap/" class="md-nav__link">


  <span class="md-ellipsis">
    Parametric UMAP
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="mutual_nn_umap/" class="md-nav__link">


  <span class="md-ellipsis">
    Mutual Nearest Neighbors UMAP
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="transform/" class="md-nav__link">


  <span class="md-ellipsis">
    Transform & Inference
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="transform_landmarked_pumap/" class="md-nav__link">


  <span class="md-ellipsis">
    Landmarked Parametric UMAP
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >


          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">


  <span class="md-ellipsis">
    Visualization and Interactive
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Visualization and Interactive
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="plotting/" class="md-nav__link">


  <span class="md-ellipsis">
    Plotting
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="interactive_viz/" class="md-nav__link">


  <span class="md-ellipsis">
    Interactive Visualization
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="outliers/" class="md-nav__link">


  <span class="md-ellipsis">
    Outlier Detection
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="document_embedding/" class="md-nav__link">


  <span class="md-ellipsis">
    Document Embedding
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="nomic_atlas_umap_of_text_embeddings/" class="md-nav__link">


  <span class="md-ellipsis">
    Nomic Atlas - Text Embeddings
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="nomic_atlas_visualizing_mnist_training_dynamics/" class="md-nav__link">


  <span class="md-ellipsis">
    Nomic Atlas - MNIST Training
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_7" >


          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">


  <span class="md-ellipsis">
    Performance & Optimization
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Performance & Optimization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="performance/" class="md-nav__link">


  <span class="md-ellipsis">
    Performance
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="benchmarking/" class="md-nav__link">


  <span class="md-ellipsis">
    Benchmarking
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="precomputed_k-nn/" class="md-nav__link">


  <span class="md-ellipsis">
    Precomputed k-NN
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>













    <li class="md-nav__item md-nav__item--nested">





        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_8" >


          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">


  <span class="md-ellipsis">
    Reference & Support
  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Reference & Support
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="api/" class="md-nav__link">


  <span class="md-ellipsis">
    API Reference
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="faq/" class="md-nav__link">


  <span class="md-ellipsis">
    FAQ
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="scientific_papers/" class="md-nav__link">


  <span class="md-ellipsis">
    Scientific Papers
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="release_notes/" class="md-nav__link">


  <span class="md-ellipsis">
    Release Notes
  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="development_roadmap/" class="md-nav__link">


  <span class="md-ellipsis">
    Development Roadmap
  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>



  </ul>
</nav>
                  </div>
                </div>
              </div>



              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">


<nav class="md-nav md-nav--secondary" aria-label="Table of contents">






    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>

        <li class="md-nav__item">
  <a href="#installing" class="md-nav__link">
    <span class="md-ellipsis">
      Installing
    </span>
  </a>

    <nav class="md-nav" aria-label="Installing">
      <ul class="md-nav__list">

          <li class="md-nav__item">
  <a href="#install-options" class="md-nav__link">
    <span class="md-ellipsis">
      Install Options
    </span>
  </a>

</li>

          <li class="md-nav__item">
  <a href="#manual-development-install" class="md-nav__link">
    <span class="md-ellipsis">
      Manual Development Install
    </span>
  </a>

</li>

          <li class="md-nav__item">
  <a href="#development-testing-with-testmon" class="md-nav__link">
    <span class="md-ellipsis">
      Development Testing with testmon
    </span>
  </a>

</li>

      </ul>
    </nav>

</li>

        <li class="md-nav__item">
  <a href="#how-to-use-umap" class="md-nav__link">
    <span class="md-ellipsis">
      How to use UMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#benefits-of-umap" class="md-nav__link">
    <span class="md-ellipsis">
      Benefits of UMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#performance-and-examples" class="md-nav__link">
    <span class="md-ellipsis">
      Performance and Examples
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#plotting" class="md-nav__link">
    <span class="md-ellipsis">
      Plotting
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#parametric-umap" class="md-nav__link">
    <span class="md-ellipsis">
      Parametric UMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#densmap" class="md-nav__link">
    <span class="md-ellipsis">
      densMAP
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#interactive-umap-with-nomic-atlas" class="md-nav__link">
    <span class="md-ellipsis">
      Interactive UMAP with Nomic Atlas
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#gpu-accelerated-umap-with-torchdr" class="md-nav__link">
    <span class="md-ellipsis">
      GPU-Accelerated UMAP with torchdr
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#help-and-support" class="md-nav__link">
    <span class="md-ellipsis">
      Help and Support
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      Citation
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    <span class="md-ellipsis">
      License
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#contributing" class="md-nav__link">
    <span class="md-ellipsis">
      Contributing
    </span>
  </a>

</li>

    </ul>

</nav>
                  </div>
                </div>
              </div>



            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">







<h1 id="umap">UMAP<a class="headerlink" href="#umap" title="Permanent link">&para;</a></h1>
<p><img alt="UMAP logo" src="doc/logo_large.png" /></p>
<p><a href="https://pypi.python.org/pypi/umap/"><img alt="PyPI Version" src="https://img.shields.io/pypi/v/umap.svg" /></a>
<a href="https://pepy.tech/project/umap"><img alt="PyPI Downloads" src="https://pepy.tech/badge/umap/month" /></a>
<a href="https://github.com/lmcinnes/umap/blob/master/LICENSE.txt"><img alt="License" src="https://img.shields.io/pypi/l/umap.svg" /></a>
<a href="https://dev.azure.com/TutteInstitute/build-pipelines/_build/latest?definitionId=2&amp;branchName=master"><img alt="Build Status" src="https://dev.azure.com/TutteInstitute/build-pipelines/_apis/build/status/lmcinnes.umap?branchName=master" /></a>
<a href="https://coveralls.io/github/lmcinnes/umap"><img alt="Coverage" src="https://coveralls.io/repos/github/lmcinnes/umap/badge.svg" /></a>
<a href="https://umap.readthedocs.io/en/latest/?badge=latest"><img alt="Docs" src="https://readthedocs.org/projects/umap/badge/?version=latest" /></a>
<a href="https://doi.org/10.21105/joss.00861"><img alt="JOSS Paper" src="http://joss.theoj.org/papers/10.21105/joss.00861/status.svg" /></a></p>
<blockquote>
<p><strong>This is a fork of the original UMAP repository</strong> (<a href="https://github.com/lmcinnes/umap">lmcinnes/umap</a>).
For the official UMAP project, please visit the original repository.</p>
</blockquote>
<p>Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique that can be used for visualisation similarly to t-SNE, but also for general non-linear dimension reduction. The algorithm is founded on three assumptions about the data:</p>
<ol>
<li>The data is uniformly distributed on a Riemannian manifold;</li>
<li>The Riemannian metric is locally constant (or can be approximated as such);</li>
<li>The manifold is locally connected.</li>
</ol>
<p>From these assumptions it is possible to model the manifold with a fuzzy topological structure. The embedding is found by searching for a low dimensional projection of the data that has the closest possible equivalent fuzzy topological structure.</p>
<p>The details for the underlying mathematics can be found in <a href="https://arxiv.org/abs/1802.03426">our paper on ArXiv</a>:</p>
<p>McInnes, L, Healy, J, <em>UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction</em>, ArXiv e-prints 1802.03426, 2018</p>
<p>A broader introduction to UMAP targetted the scientific community can be found in our <a href="https://doi.org/10.1038/s43586-024-00363-x">paper published in Nature Review Methods Primers</a>:</p>
<p>Healy, J., McInnes, L. <em>Uniform manifold approximation and projection</em>. Nat Rev Methods Primers 4, 82 (2024).</p>
<p>A read only version of this paper can accessed via <a href="https://rdcu.be/d0YZT">link</a></p>
<p>The important thing is that you don't need to worry about that—you can use UMAP right now for dimension reduction and visualisation as easily as a drop in replacement for scikit-learn's t-SNE.</p>
<p>Documentation is <a href="https://umap.readthedocs.io/">available via Read the Docs</a>.</p>
<p><strong>New: this package now also provides support for densMAP.</strong> The densMAP algorithm augments UMAP to preserve local density information in addition to the topological structure of the data. Details of this method are described in the following <a href="https://doi.org/10.1038/s41587-020-00801-7">paper</a>:</p>
<p>Narayan, A, Berger, B, Cho, H, <em>Assessing Single-Cell Transcriptomic Variability through Density-Preserving Data Visualization</em>, Nature Biotechnology, 2021</p>
<h2 id="installing">Installing<a class="headerlink" href="#installing" title="Permanent link">&para;</a></h2>
<p>UMAP depends upon <code>scikit-learn</code>, and thus <code>scikit-learn</code>'s dependencies such as <code>numpy</code> and <code>scipy</code>. UMAP adds a requirement for <code>numba</code> for performance reasons. The original version used Cython, but the improved code clarity, simplicity and performance of Numba made the transition necessary.</p>
<p><strong>New in this version:</strong> This fork includes an optimized Rust-based HNSW nearest neighbor backend (<code>hnsw-rs</code>) that provides significant performance improvements over the default implementations, especially for large datasets.</p>
<p><strong>Requirements:</strong></p>
<ul>
<li>Python 3.6 or greater</li>
<li>numpy</li>
<li>scipy</li>
<li>scikit-learn</li>
<li>numba</li>
<li>tqdm</li>
<li><a href="https://github.com/lmcinnes/pynndescent">pynndescent</a></li>
</ul>
<p><strong>Recommended packages:</strong></p>
<ul>
<li>For plotting</li>
<li>matplotlib</li>
<li>datashader</li>
<li>holoviews</li>
<li>For Parametric UMAP (PyTorch-based)</li>
<li>torch &gt;= 1.9.0</li>
<li>torchvision (for image utilities)</li>
<li>For optimized nearest neighbor search</li>
<li>hnsw-rs (Rust-based HNSW backend)</li>
</ul>
<h3 id="install-options">Install Options<a class="headerlink" href="#install-options" title="Permanent link">&para;</a></h3>
<p>The recommended way to install UMAP is via PyPI using uv, which provides faster and more reliable dependency resolution:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>uv<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>umap
</code></pre></div>
<p>If you don't have uv installed, you can install it from <a href="https://docs.astral.sh/uv/getting-started/installation/">https://docs.astral.sh/uv/getting-started/installation/</a>.</p>
<p>Alternatively, you can install UMAP using pip:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>umap
</code></pre></div>
<p>This will install UMAP and all required dependencies.</p>
<p>If you wish to use the plotting functionality you can use</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>uv<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>umap<span class="o">[</span>plot<span class="o">]</span>
</code></pre></div>
<p>or with pip:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>umap<span class="o">[</span>plot<span class="o">]</span>
</code></pre></div>
<p>to install all the plotting dependencies.</p>
<p>If you wish to use Parametric UMAP, you need to install PyTorch, which can be installed using the instructions at <a href="https://pytorch.org/get-started/locally">https://pytorch.org/get-started/locally</a> (choose your platform and preferences), or use:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>uv<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>umap<span class="o">[</span>parametric_umap<span class="o">]</span>
</code></pre></div>
<p>or with pip:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>umap<span class="o">[</span>parametric_umap<span class="o">]</span>
</code></pre></div>
<p>This will install PyTorch with CPU support. For GPU support, follow the <a href="https://pytorch.org/get-started/locally">official PyTorch installation guide</a> to install the appropriate CUDA version.</p>
<p>If you're on an x86 processor, you can also optionally install <code>tbb</code>, which will provide additional CPU optimizations:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>uv<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>umap<span class="o">[</span>tbb<span class="o">]</span>
</code></pre></div>
<p>or with pip:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span>umap<span class="o">[</span>tbb<span class="o">]</span>
</code></pre></div>
<h3 id="manual-development-install">Manual Development Install<a class="headerlink" href="#manual-development-install" title="Permanent link">&para;</a></h3>
<p>For a manual development install, clone the repository and install in editable mode:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/lmcinnes/umap.git
<span class="nb">cd</span><span class="w"> </span>umap
</code></pre></div>
<p>Then create a virtual environment (recommended) and install the package using uv:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>uv<span class="w"> </span>venv<span class="w"> </span>venv
<span class="nb">source</span><span class="w"> </span>venv/bin/activate<span class="w">  </span><span class="c1"># On Windows: venv\Scripts\activate</span>
uv<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>.
</code></pre></div>
<p>Or if you prefer to use pip:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code>python<span class="w"> </span>-m<span class="w"> </span>venv<span class="w"> </span>venv
<span class="nb">source</span><span class="w"> </span>venv/bin/activate<span class="w">  </span><span class="c1"># On Windows: venv\Scripts\activate</span>
pip<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>.
</code></pre></div>
<h3 id="development-testing-with-testmon">Development Testing with testmon<a class="headerlink" href="#development-testing-with-testmon" title="Permanent link">&para;</a></h3>
<p>This project uses <code>pytest-testmon</code> to optimize test execution by only running tests affected by code changes. For best results (especially in CI/CD environments), ensure testmon uses a file-based database:</p>
<div class="highlight"><span class="filename">Bash</span><pre><span></span><code><span class="c1"># Run tests with file-based testmon cache</span>
pytest<span class="w"> </span>--testmon-db<span class="o">=</span>.testmon.db<span class="w"> </span>umap/tests/
</code></pre></div>
<p>Configure in <code>pyproject.toml</code> or pytest config:</p>
<div class="highlight"><span class="filename">INI</span><pre><span></span><code><span class="k">[tool:pytest]</span>
<span class="na">testmon_db</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">.testmon.db</span>
</code></pre></div>
<p>The file-based database (<code>.testmon.db</code>) should be persistent across test runs to provide reliable optimization benefits.</p>
<h2 id="how-to-use-umap">How to use UMAP<a class="headerlink" href="#how-to-use-umap" title="Permanent link">&para;</a></h2>
<p>The umap package inherits from sklearn classes, and thus drops in neatly next to other sklearn transformers with an identical calling API.</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="kn">import</span> <span class="nn">umap</span>
<span class="kn">from</span> <span class="nn">sklearn.datasets</span> <span class="kn">import</span> <span class="n">load_digits</span>

<span class="n">digits</span> <span class="o">=</span> <span class="n">load_digits</span><span class="p">()</span>

<span class="n">embedding</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">digits</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>
<p>There are a number of parameters that can be set for the UMAP class; the major ones are as follows:</p>
<ul>
<li>
<p><code>n_neighbors</code>: This determines the number of neighboring points used in local approximations of manifold structure. Larger values will result in more global structure being preserved at the loss of detailed local structure. In general this parameter should often be in the range 5 to 50, with a choice of 10 to 15 being a sensible default.</p>
</li>
<li>
<p><code>min_dist</code>: This controls how tightly the embedding is allowed compress points together. Larger values ensure embedded points are more evenly distributed, while smaller values allow the algorithm to optimise more accurately with regard to local structure. Sensible values are in the range 0.001 to 0.5, with 0.1 being a reasonable default.</p>
</li>
<li>
<p><code>metric</code>: This determines the choice of metric used to measure distance in the input space. A wide variety of metrics are already coded, and a user defined function can be passed as long as it has been JITd by numba.</p>
</li>
</ul>
<p>An example of making use of these options:</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="kn">import</span> <span class="nn">umap</span>
<span class="kn">from</span> <span class="nn">sklearn.datasets</span> <span class="kn">import</span> <span class="n">load_digits</span>

<span class="n">digits</span> <span class="o">=</span> <span class="n">load_digits</span><span class="p">()</span>

<span class="n">embedding</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                      <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span>
                      <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">digits</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>
<p>UMAP also supports fitting to sparse matrix data. For more details please see <a href="https://umap.readthedocs.io/">the UMAP documentation</a></p>
<h2 id="benefits-of-umap">Benefits of UMAP<a class="headerlink" href="#benefits-of-umap" title="Permanent link">&para;</a></h2>
<p>UMAP has a few signficant wins in its current incarnation.</p>
<p>First of all UMAP is <em>fast</em>. It can handle large datasets and high dimensional data without too much difficulty, scaling beyond what most t-SNE packages can manage. This includes very high dimensional sparse datasets. UMAP has successfully been used directly on data with over a million dimensions.</p>
<p>Second, UMAP scales well in embedding dimension—it isn't just for visualisation! You can use UMAP as a general purpose dimension reduction technique as a preliminary step to other machine learning tasks. With a little care it partners well with the <a href="https://github.com/scikit-learn-contrib/hdbscan">hdbscan</a> clustering library (for more details please see <a href="https://umap.readthedocs.io/en/latest/clustering.html">Using UMAP for Clustering</a>).</p>
<p>Third, UMAP often performs better at preserving some aspects of global structure of the data than most implementations of t-SNE. This means that it can often provide a better "big picture" view of your data as well as preserving local neighbor relations.</p>
<p>Fourth, UMAP supports a wide variety of distance functions, including non-metric distance functions such as <em>cosine distance</em> and <em>correlation distance</em>. You can finally embed word vectors properly using cosine distance!</p>
<p>Fifth, UMAP supports adding new points to an existing embedding via the standard sklearn <code>transform</code> method. This means that UMAP can be used as a preprocessing transformer in sklearn pipelines.</p>
<p>Sixth, UMAP supports supervised and semi-supervised dimension reduction. This means that if you have label information that you wish to use as extra information for dimension reduction (even if it is just partial labelling) you can do that—as simply as providing it as the <code>y</code> parameter in the fit method.</p>
<p>Seventh, UMAP supports a variety of additional experimental features including: an "inverse transform" that can approximate a high dimensional sample that would map to a given position in the embedding space; the ability to embed into non-euclidean spaces including hyperbolic embeddings, and embeddings with uncertainty; very preliminary support for embedding dataframes also exists.</p>
<p>Finally, UMAP has solid theoretical foundations in manifold learning (see <a href="https://arxiv.org/abs/1802.03426">our paper on ArXiv</a>). This both justifies the approach and allows for further extensions that will soon be added to the library.</p>
<h2 id="performance-and-examples">Performance and Examples<a class="headerlink" href="#performance-and-examples" title="Permanent link">&para;</a></h2>
<p>UMAP is very efficient at embedding large high dimensional datasets. In particular it scales well with both input dimension and embedding dimension. For the best possible performance we recommend:</p>
<ol>
<li>Installing the Rust-based HNSW backend (<a href="https://github.com/lmcinnes/hnsw-rs">hnsw-rs</a>) for optimized nearest neighbor search</li>
<li>Installing the nearest neighbor computation library <a href="https://github.com/lmcinnes/pynndescent">pynndescent</a> as a fallback</li>
</ol>
<p>UMAP will work without these packages, but with them installed it will run significantly faster, particularly on multicore machines and large datasets.</p>
<p>For a problem such as the 784-dimensional MNIST digits dataset with 70000 data samples, UMAP can complete the embedding in under a minute (as compared with around 45 minutes for scikit-learn's t-SNE implementation). Despite this runtime efficiency, UMAP still produces high quality embeddings.</p>
<p>The obligatory MNIST digits dataset, embedded in 42 seconds (with pynndescent installed and after numba jit warmup) using a 3.1 GHz Intel Core i7 processor (n_neighbors=10, min_dist=0.001):</p>
<p><img alt="UMAP embedding of MNIST digits" src="images/umap_example_mnist1.png" /></p>
<p>The MNIST digits dataset is fairly straightforward, however. A better test is the more recent "Fashion MNIST" dataset of images of fashion items (again 70000 data sample in 784 dimensions). UMAP produced this embedding in 49 seconds (n_neighbors=5, min_dist=0.1):</p>
<p><img alt="UMAP embedding of &quot;Fashion MNIST&quot;" src="images/umap_example_fashion_mnist1.png" /></p>
<p>The UCI shuttle dataset (43500 sample in 8 dimensions) embeds well under <em>correlation</em> distance in 44 seconds (note the longer time required for correlation distance computations):</p>
<p><img alt="UMAP embedding the UCI Shuttle dataset" src="images/umap_example_shuttle.png" /></p>
<p>The following is a densMAP visualization of the MNIST digits dataset with 784 features based on the same parameters as above (n_neighbors=10, min_dist=0.001). densMAP reveals that the cluster corresponding to digit 1 is noticeably denser, suggesting that there are fewer degrees of freedom in the images of 1 compared to other digits.</p>
<p><img alt="densMAP embedding of the MNIST dataset" src="images/densmap_example_mnist.png" /></p>
<h2 id="plotting">Plotting<a class="headerlink" href="#plotting" title="Permanent link">&para;</a></h2>
<p>UMAP includes a subpackage <code>umap.plot</code> for plotting the results of UMAP embeddings. This package needs to be imported separately since it has extra requirements (matplotlib, datashader and holoviews). It allows for fast and simple plotting and attempts to make sensible decisions to avoid overplotting and other pitfalls. An example of use:</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="kn">import</span> <span class="nn">umap</span>
<span class="kn">import</span> <span class="nn">umap.plot</span>
<span class="kn">from</span> <span class="nn">sklearn.datasets</span> <span class="kn">import</span> <span class="n">load_digits</span>

<span class="n">digits</span> <span class="o">=</span> <span class="n">load_digits</span><span class="p">()</span>

<span class="n">mapper</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">digits</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="n">umap</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">points</span><span class="p">(</span><span class="n">mapper</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">digits</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
</code></pre></div>
<p>The plotting package offers basic plots, as well as interactive plots with hover tools and various diagnostic plotting options. See the documentation for more details.</p>
<h2 id="parametric-umap">Parametric UMAP<a class="headerlink" href="#parametric-umap" title="Permanent link">&para;</a></h2>
<p>Parametric UMAP provides support for training a neural network to learn a UMAP based transformation of data using PyTorch. This can be used to support faster inference of new unseen data, more robust inverse transforms, autoencoder versions of UMAP and semi-supervised classification (particularly for data well separated by UMAP and very limited amounts of labelled data). The PyTorch-based implementation provides better performance and flexibility compared to TensorFlow. See the <a href="https://umap.readthedocs.io/en/0.5dev/parametric_umap.html">documentation of Parametric UMAP</a> or the <a href="https://github.com/lmcinnes/umap/tree/master/notebooks/Parametric_UMAP">example notebooks</a> for more.</p>
<h2 id="densmap">densMAP<a class="headerlink" href="#densmap" title="Permanent link">&para;</a></h2>
<p>The densMAP algorithm augments UMAP to additionally preserve local density information in addition to the topological structure captured by UMAP. One can easily run densMAP using the umap package by setting the <code>densmap</code> input flag:</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="n">embedding</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">densmap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>
<p>This functionality is built upon the densMAP <a href="https://github.com/hhcho/densvis">implementation</a> provided by the developers of densMAP, who also contributed to integrating densMAP into the umap package.</p>
<p>densMAP inherits all of the parameters of UMAP. The following is a list of additional parameters that can be set for densMAP:</p>
<ul>
<li>
<p><code>dens_frac</code>: This determines the fraction of epochs (a value between 0 and 1) that will include the density-preservation term in the optimization objective. This parameter is set to 0.3 by default. Note that densMAP switches density optimization on after an initial phase of optimizing the embedding using UMAP.</p>
</li>
<li>
<p><code>dens_lambda</code>: This determines the weight of the density-preservation objective. Higher values prioritize density preservation, and lower values (closer to zero) prioritize the UMAP objective. Setting this parameter to zero reduces the algorithm to UMAP. Default value is 2.0.</p>
</li>
<li>
<p><code>dens_var_shift</code>: Regularization term added to the variance of local densities in the embedding for numerical stability. We recommend setting this parameter to 0.1, which consistently works well in many settings.</p>
</li>
<li>
<p><code>output_dens</code>: When this flag is True, the call to <code>fit_transform</code> returns, in addition to the embedding, the local radii (inverse measure of local density defined in the <a href="https://doi.org/10.1101/2020.05.12.077776">densMAP paper</a>) for the original dataset and for the embedding. The output is a tuple <code>(embedding, radii_original, radii_embedding)</code>. Note that the radii are log-transformed. If False, only the embedding is returned. This flag can also be used with UMAP to explore the local densities of UMAP embeddings. By default this flag is False.</p>
</li>
</ul>
<p>For densMAP we recommend larger values of <code>n_neighbors</code> (e.g. 30) for reliable estimation of local density.</p>
<p>An example of making use of these options (based on a subsample of the mnist_784 dataset):</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="kn">import</span> <span class="nn">umap</span>
<span class="kn">from</span> <span class="nn">sklearn.datasets</span> <span class="kn">import</span> <span class="n">fetch_openml</span>
<span class="kn">from</span> <span class="nn">sklearn.utils</span> <span class="kn">import</span> <span class="n">resample</span>

<span class="n">digits</span> <span class="o">=</span> <span class="n">fetch_openml</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;mnist_784&#39;</span><span class="p">)</span>
<span class="n">subsample</span><span class="p">,</span> <span class="n">subsample_labels</span> <span class="o">=</span> <span class="n">resample</span><span class="p">(</span><span class="n">digits</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">digits</span><span class="o">.</span><span class="n">target</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="mi">7000</span><span class="p">,</span>
                                       <span class="n">stratify</span><span class="o">=</span><span class="n">digits</span><span class="o">.</span><span class="n">target</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">embedding</span><span class="p">,</span> <span class="n">r_orig</span><span class="p">,</span> <span class="n">r_emb</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">densmap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dens_lambda</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
                                     <span class="n">output_dens</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">subsample</span><span class="p">)</span>
</code></pre></div>
<p>See <a href="https://umap.readthedocs.io/en/0.5dev/densmap_demo.html">the documentation</a> for more details.</p>
<h2 id="interactive-umap-with-nomic-atlas">Interactive UMAP with Nomic Atlas<a class="headerlink" href="#interactive-umap-with-nomic-atlas" title="Permanent link">&para;</a></h2>
<p><img alt="MNIST UMAP visualization in Nomic Atlas" src="https://assets.nomicatlas.com/mnist-training-embeddings-umap-short.gif" /></p>
<p>For interactive exploration of UMAP embeddings, especially for visualizing large datasets data over time/training epochs, you can use <a href="https://atlas.nomic.ai/">Nomic Atlas</a>. Nomic Atlas is a platform for embedding generation, visualization, analysis, and retrieval that directly integrates UMAP as one of its projection models.</p>
<p>Using Nomic Atlas with UMAP is straightforward:</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="kn">from</span> <span class="nn">nomic</span> <span class="kn">import</span> <span class="n">AtlasDataset</span>
<span class="kn">from</span> <span class="nn">nomic.data_inference</span> <span class="kn">import</span> <span class="n">ProjectionOptions</span>

<span class="c1"># Create a dataset</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">AtlasDataset</span><span class="p">(</span><span class="s2">&quot;my-dataset&quot;</span><span class="p">)</span>

<span class="c1"># data is a DataFrame or a list of dicts</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">add_data</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="c1"># Create an interactive UMAP in Atlas</span>
<span class="n">atlas_map</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">create_index</span><span class="p">(</span>
    <span class="n">indexed_field</span><span class="o">=</span><span class="s1">&#39;text&#39;</span><span class="p">,</span>
    <span class="n">projection</span><span class="o">=</span><span class="n">ProjectionOptions</span><span class="p">(</span>
        <span class="n">model</span><span class="o">=</span><span class="s2">&quot;umap&quot;</span><span class="p">,</span>
        <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
        <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="n">n_epochs</span><span class="o">=</span><span class="mi">200</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="c1"># you can access your UMAP coordinates later on with</span>
<span class="c1"># atlas_map.maps[0].embeddings.projected</span>
</code></pre></div>
<p>Nomic Atlas provides:</p>
<ul>
<li>In-browser analysis of your UMAP data with the <a href="https://docs.nomic.ai/atlas/data-maps/atlas-analyst">Atlas Analyst</a></li>
<li>Vector search over your UMAP data using the <a href="https://docs.nomic.ai/atlas/data-maps/guides/vector-search-over-your-data">Nomic API</a></li>
<li>Interactive features like zooming, recoloring, searching, and filtering in the <a href="https://docs.nomic.ai/atlas/data-maps/controls">Nomic Atlas data map</a></li>
<li>Scalability for millions of data points</li>
<li>Rich information display on hover</li>
<li>Shareable UMAPs via URL links to your embeddings and data maps in Atlas</li>
</ul>
<h2 id="gpu-accelerated-umap-with-torchdr">GPU-Accelerated UMAP with torchdr<a class="headerlink" href="#gpu-accelerated-umap-with-torchdr" title="Permanent link">&para;</a></h2>
<p>For GPU-accelerated UMAP computations, <a href="https://github.com/TorchDR/TorchDR">torchdr</a> provides a PyTorch-based implementation that significantly speed up the algorithm. torchdr accelerates <strong>every step</strong> of the dimensionality reduction pipeline on GPU: kNN computation, affinity construction and embedding optimization.</p>
<p>Using torchdr with UMAP is straightforward:</p>
<div class="highlight"><span class="filename">Python</span><pre><span></span><code><span class="kn">from</span> <span class="nn">torchdr</span> <span class="kn">import</span> <span class="n">UMAP</span> <span class="k">as</span> <span class="n">torchdrUMAP</span>

<span class="n">umap_gpu</span> <span class="o">=</span> <span class="n">torchdrUMAP</span><span class="p">(</span>
    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda&#39;</span>
<span class="p">)</span>
<span class="n">embedding</span> <span class="o">=</span> <span class="n">umap_gpu</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">data</span><span class="o">-</span><span class="n">maps</span><span class="p">)</span>
</code></pre></div>
<p>For more information and advanced usage, see the <a href="https://torchdr.github.io/index.html">torchdr documentation</a>.</p>
<h2 id="help-and-support">Help and Support<a class="headerlink" href="#help-and-support" title="Permanent link">&para;</a></h2>
<p>Documentation is at <a href="https://umap.readthedocs.io/">Read the Docs</a>. The documentation <a href="https://umap.readthedocs.io/en/latest/faq.html">includes a FAQ</a> that may answer your questions. If you still have questions then please <a href="https://github.com/lmcinnes/umap/issues/new">open an issue</a> and I will try to provide any help and guidance that I can.</p>
<h2 id="citation">Citation<a class="headerlink" href="#citation" title="Permanent link">&para;</a></h2>
<p>If you make use of this software for your work we would appreciate it if you would cite the paper from the Journal of Open Source Software:</p>
<div class="highlight"><span class="filename">BibTeX</span><pre><span></span><code><span class="nc">@article</span><span class="p">{</span><span class="nl">mcinnes2018umap-software</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="p">=</span><span class="s">{UMAP: Uniform Manifold Approximation and Projection}</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="p">=</span><span class="s">{McInnes, Leland and Healy, John and Saul, Nathaniel and Grossberger, Lukas}</span><span class="p">,</span>
<span class="w">  </span><span class="na">journal</span><span class="p">=</span><span class="s">{The Journal of Open Source Software}</span><span class="p">,</span>
<span class="w">  </span><span class="na">volume</span><span class="p">=</span><span class="s">{3}</span><span class="p">,</span>
<span class="w">  </span><span class="na">number</span><span class="p">=</span><span class="s">{29}</span><span class="p">,</span>
<span class="w">  </span><span class="na">pages</span><span class="p">=</span><span class="s">{861}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="p">=</span><span class="s">{2018}</span>
<span class="p">}</span>
</code></pre></div>
<p>If you would like to cite this algorithm in your work the ArXiv paper is the current reference:</p>
<div class="highlight"><span class="filename">BibTeX</span><pre><span></span><code><span class="nc">@article</span><span class="p">{</span><span class="nl">2018arXivUMAP</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{{McInnes}, L. and {Healy}, J. and {Melville}, J.}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;{UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction}&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="na">journal</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{ArXiv e-prints}</span><span class="p">,</span>
<span class="w">  </span><span class="na">archivePrefix</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;arXiv&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="na">eprint</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{1802.03426}</span><span class="p">,</span>
<span class="w">  </span><span class="na">primaryClass</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;stat.ML&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="na">keywords</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Statistics - Machine Learning, Computer Science - Computational Geometry, Computer Science - Learning}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="m">2018</span><span class="p">,</span>
<span class="w">  </span><span class="na">month</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="nv">feb</span><span class="p">,</span>
<span class="p">}</span>
</code></pre></div>
<p>If you found the Nature Primer introduction useful please cite the following reference:</p>
<div class="highlight"><span class="filename">BibTeX</span><pre><span></span><code><span class="nc">@article</span><span class="p">{</span><span class="nl">Healy2024</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="p">=</span><span class="s">{Healy, John and McInnes, Leland}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="p">=</span><span class="s">{Uniform manifold approximation and projection}</span><span class="p">,</span>
<span class="w">  </span><span class="na">journal</span><span class="p">=</span><span class="s">{Nature Reviews Methods Primers}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="p">=</span><span class="s">{2024}</span><span class="p">,</span>
<span class="w">  </span><span class="na">month</span><span class="p">=</span><span class="s">{Nov}</span><span class="p">,</span>
<span class="w">  </span><span class="na">day</span><span class="p">=</span><span class="s">{21}</span><span class="p">,</span>
<span class="w">  </span><span class="na">volume</span><span class="p">=</span><span class="s">{4}</span><span class="p">,</span>
<span class="w">  </span><span class="na">number</span><span class="p">=</span><span class="s">{1}</span><span class="p">,</span>
<span class="w">  </span><span class="na">pages</span><span class="p">=</span><span class="s">{82}</span><span class="p">,</span>
<span class="w">  </span><span class="na">abstract</span><span class="p">=</span><span class="s">{Uniform manifold approximation and projection is a nonlinear dimension reduction method often used for visualizing data and as pre-processing for further machine-learning tasks such as clustering. In this Primer, we provide an introduction to the uniform manifold approximation and projection algorithm, the intuitions behind how it works, how best to apply it on data and how to interpret and understand results.}</span><span class="p">,</span>
<span class="w">  </span><span class="na">issn</span><span class="p">=</span><span class="s">{2662-8449}</span><span class="p">,</span>
<span class="w">  </span><span class="na">doi</span><span class="p">=</span><span class="s">{10.1038/s43586-024-00363-x}</span><span class="p">,</span>
<span class="w">  </span><span class="na">url</span><span class="p">=</span><span class="s">{https://doi.org/10.1038/s43586-024-00363-x}</span>
<span class="p">}</span>
</code></pre></div>
<p>Additionally, if you use the densMAP algorithm in your work please cite the following reference:</p>
<div class="highlight"><span class="filename">BibTeX</span><pre><span></span><code><span class="nc">@article</span><span class="w"> </span><span class="p">{</span><span class="nl">NBC2020</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Narayan, Ashwin and Berger, Bonnie and Cho, Hyunghoon}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Assessing Single-Cell Transcriptomic Variability through Density-Preserving Data Visualization}</span><span class="p">,</span>
<span class="w">  </span><span class="na">journal</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Nature Biotechnology}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2021}</span><span class="p">,</span>
<span class="w">  </span><span class="na">doi</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{10.1038/s41587-020-00801-7}</span><span class="p">,</span>
<span class="w">  </span><span class="na">publisher</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Springer Nature}</span><span class="p">,</span>
<span class="w">  </span><span class="na">URL</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://doi.org/10.1038/s41587-020-00801-7}</span><span class="p">,</span>
<span class="w">  </span><span class="na">eprint</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://www.biorxiv.org/content/early/2020/05/14/2020.05.12.077776.full.pdf}</span><span class="p">,</span>
<span class="p">}</span>
</code></pre></div>
<p>If you use the Parametric UMAP algorithm in your work please cite the following reference:</p>
<div class="highlight"><span class="filename">BibTeX</span><pre><span></span><code><span class="nc">@article</span><span class="w"> </span><span class="p">{</span><span class="nl">SMG2020</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Sainburg, Tim and McInnes, Leland and Gentner, Timothy Q.}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Parametric UMAP: learning embeddings with deep neural networks for representation and semi-supervised learning}</span><span class="p">,</span>
<span class="w">  </span><span class="na">journal</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{ArXiv e-prints}</span><span class="p">,</span>
<span class="w">  </span><span class="na">archivePrefix</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;arXiv&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="na">eprint</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2009.12981}</span><span class="p">,</span>
<span class="w">  </span><span class="na">primaryClass</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">&quot;stat.ML&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="na">keywords</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Statistics - Machine Learning, Computer Science - Computational Geometry, Computer Science - Learning}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="m">2020</span><span class="p">,</span>
<span class="p">}</span>
</code></pre></div>
<h2 id="license">License<a class="headerlink" href="#license" title="Permanent link">&para;</a></h2>
<p>The umap package is 3-clause BSD licensed.</p>
<p>We would like to note that the umap package makes heavy use of NumFOCUS sponsored projects, and would not be possible without their support of those projects, so please <a href="https://www.numfocus.org/membership">consider contributing to NumFOCUS</a>.</p>
<h2 id="contributing">Contributing<a class="headerlink" href="#contributing" title="Permanent link">&para;</a></h2>
<p>Contributions are more than welcome! There are lots of opportunities for potential projects, so please get in touch if you would like to help out. Everything from code to notebooks to examples and documentation are all <em>equally valuable</em> so please don't feel you can't contribute. To contribute please <a href="https://github.com/lmcinnes/umap/issues#fork-destination-box">fork the project</a> make your changes and submit a pull request. We will do our best to work through any issues with you and get your code merged into the main branch.</p>













              </article>
            </div>


<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>

          <button type="button" class="md-top md-icon" data-md-component="top" hidden>

  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>

      </main>

        <footer class="md-footer">

  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">

    <div class="md-copyright__highlight">
      Copyright &copy; 2017-2024 Leland McInnes
    </div>


    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>

</div>

    </div>
  </div>
</footer>

    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>


    <script id="__config" type="application/json">{"base": ".", "features": ["navigation.instant", "navigation.expand", "navigation.top", "content.code.copy", "content.tooltips", "search.suggest", "search.highlight"], "search": "assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>


      <script src="assets/javascripts/bundle.83f73b43.min.js"></script>


  </body>
</html>
